conngen/conngen.cpp

```diff
@@ -37,185 +34,312 @@
+    np, ConnectionGenerator::Mask( 1, np ) );
 
-  const modelrange source_range =
-    kernel().modelrange_manager.get_contiguous_gid_range(
-      ( *sources->local_begin() )->get_gid() );
-  index source_offset = source_range.get_first_gid();
-  RangeSet source_ranges;
-  source_ranges.push_back(
-    Range( source_range.get_first_gid(), source_range.get_last_gid() ) );
-
-  const modelrange target_range =
-    kernel().modelrange_manager.get_contiguous_gid_range(
-      ( *targets->local_begin() )->get_gid() );
-  index target_offset = target_range.get_first_gid();
-  RangeSet target_ranges;
-  target_ranges.push_back(
-    Range( target_range.get_first_gid(), target_range.get_last_gid() ) );
-
-  cg_connect( cg,
-    source_ranges,
-    source_offset,
-    target_ranges,
-    target_offset,
-    params_map,
-    synmodel_id );
+  cg_create_masks( &masks, sources, targets );
+  cg->setMask( masks, kernel().mpi_manager.get_rank() );
 }
 
+/**
+ * Create the masks for sources and targets based on the contiguous
+ * ranges given in sources and targets. We need to do some index
+ * translation here, as the CG expects indices from 0..n for both
+ * source and target populations, while the RangeSets sources and
+ * targets contain NEST global indices (gids).
+ *
+ * The masks for the sources must contain all nodes (local+remote).
+ * The skip of the mask was set to 1 in cg_set_masks(). The same
+ * source mask is stored n_proc times on each process.
+ *
+ * The masks for the targets must only contain local nodes. This is
+ * achieved by first setting skip to num_processes upon creation of
+ * the mask in cg_set_masks(), and second by the fact that for each
+ * contiguous range of nodes in a mask, each of them contains the
+ * index-translated id of the first local neuron as the first
+ * entry. If this renders the range empty (i.e. because the first
+ * local id is beyond the last element of the range), the range is
+ * not added to the mask.
+ *
+ * \param masks The std::vector of Masks to populate
+ * \param sources The source ranges to create the source masks from
+ * \param targets The target ranges to create the target masks from
+ *
+ * \note Each process computes the full set of source and target
+ * masks, i.e. one mask per rank will be created on each rank.
+ *
+ * \note Setting the masks for all processes on each process might
+ * become a memory bottleneck when going to very large numbers of
+ * processes. Especially so for the source masks, which are all the
+ * same. This could be solved by making the ConnectionGenerator
+ * interface MPI aware and communicating the masks during connection
+ * setup.
+*/
 void
-nest::cg_connect( nest::ConnectionGeneratorDatum& cg,
-  IntVectorDatum& sources,
-  IntVectorDatum& targets,
-  const DictionaryDatum& params_map,
-  const Name& synmodel_name )
+cg_create_masks( std::vector< ConnectionGenerator::Mask >* masks,
+  RangeSet& sources,
+  RangeSet& targets )
```

Comment:
As far as I can see, `sources` and `targets` are not modified below. Why not pass as `const &`?

Reply:
They are `const GIDCollection&` now. See fe4408921.